### PR TITLE
Rect points

### PR DESCRIPTION
**Pull request by mwieder**
Allows getting "the effective points" of graphic rectangles, round rectangles, and regular polygons.
This allows for some things that are difficult at present: setting the points of a freeform polygon to one of several regular polygon objects; moving an image around the points of a regular polygon, etc.

### CODE REVIEW

**Comment by mwieder** [comment]
This pull request has been sitting around for a while. I just updated it to better code... any chance someone could take a look at it?

**Comment by runrevmark** [comment]
Hi Mark - didn't quite get around to looking at it today, but should tomorrow! Thanks for your patience.

**Comment by mwieder** [comment]
Thanks. Patience actually paid off because I found a better way of implementing it in the intervening months. Note: I haven't tackled the points of an oval, and I'm debating whether there's a good use case for that - the only thing I can think of is moving a graphic along the points of an oval.

**Comment by runrevmark** [comment]
Hi Mark,

I've taken an initial look at this and it's looking good. My only suggestion would be to not use realpoints/points fields in the generators for the points for things that don't have points. Instead, just generate into a temporary buffer, unparse, then delete the temporary buffer. (This would side-step sorting out a couple of memory leaks which I think might be lurking - I've marked them in the code).

Indeed, you could have a method for each of the shape-types which don't have realpoints, that generate the points array - this would nicely refactor the code in ::draw which deals with the regular polygons (which could also be done on the fly) and mean that the 'points' field in MCGraphic becomes redundent. (Time taken to generate points for a regular polygon is vastly less than the time taken to render the resulting path to the context - so there's no issue about performance here). For example:

bool MCGraphic::getpointsforrect(MCPoint_& r_points, int& r_point_count) { ... }
bool MCGraphic::getpointsforregular(MCPoint_& r_points, int& r_point_count) { ... }

Thanks again,

Mark.

File: engine/src/graphic.cpp [hunk 1]
```diff
@@ -546,14 +547,119 @@ Exec_stat MCGraphic::getprop(uint4 parid, Properties which, MCExecPoint& ep, Boo
...
+				nrealpoints = 0;
+				MCU_roundrect(realpoints, nrealpoints, geteffectiverect(), roundradius);
+				MCU_unparsepoints(realpoints, nrealpoints, ep);
+				break;
+			}
+			case F_G_RECTANGLE:
+			{
+				MCRectangle nrect = geteffectiverect();
+				nrealpoints = 4;
+				realpoints = new MCPoint[nrealpoints];
```
Comment (runrevmark) [inline engine/src/graphic.cpp]: I think there's a memory leak here as its reallocating realpoints each time the points of a rectangle are fetched.

File: engine/src/graphic.cpp [hunk 2]
```diff
@@ -546,14 +547,119 @@ Exec_stat MCGraphic::getprop(uint4 parid, Properties which, MCExecPoint& ep, Boo
 		ep.setint(nsides);
 		break;
 	case P_POINTS:
-		MCU_unparsepoints(realpoints, nrealpoints, ep);
+		// MDW-2014-06-14: [[ rect_points ]] allow effective points as read-only
+		p_flags = getstyleint(flags);
+		if (p_flags == F_ROUNDRECT || p_flags == F_G_RECTANGLE)
```
Comment (runrevmark) [inline engine/src/graphic.cpp]: This should probably include F_REGULAR too - as you can't set the points of a regular graphic as they are defined by the other properties.

(You can set the points property, if style is regular, but I don't believe it has any effect until you set the style to polygon or curve - i.e. one of the styles that uses the realpoints).

Comment (mwieder) [inline engine/src/graphic.cpp]: Quite right - I had F_REGULAR in there and then with sudden loss of clarity deleted it. I'll put it back forthwith. I don't think setting the points of rectangles, rounded rectangles, or regular polygons is a particularly good idea - there's too much validation involved to ensure that the angles and point placements are correct - that's why I enforce the "effective" modifer.

File: engine/src/graphic.cpp (hunk 1 above)
Comment (mwieder) [inline engine/src/graphic.cpp]: I think you're probably correct. The whole points/oldpoints/realpoints/markerpoints and delpoints()  thing is quite confusing and undocumented. I found that realpoints did what I wanted and stopped there. I'll see about patching the leak.

**Comment by mwieder** [comment]
OK - refactored, cleaned up the memory leaks, pulled F_REGULAR back into the fold, and created new functions to return the points of rects, roundrects, and regular polygons. The regular polygon function is now integrated into the ::draw method as well. I'm at a loss as to how to determine the points of an oval - seems like squaring the circle, but the draw method seems capable of drawing a circle through the fillarc function.

File: engine/src/graphic.cpp [hunk 3]
```diff
@@ -546,14 +550,97 @@ Exec_stat MCGraphic::getprop(uint4 parid, Properties which, MCExecPoint& ep, Boo
...
+		switch (graphic_type)
+		{
+			case F_ROUNDRECT:
+			{
+				fakepoints = NULL;
+				nfakepoints = 0;
+				get_points_for_roundrect(fakepoints, nfakepoints);
+				MCU_offset_points(fakepoints, nfakepoints, -trect.x, -trect.y);
+				MCU_unparsepoints(fakepoints, nfakepoints, ep);
+				break;
```
Comment (runrevmark) [inline engine/src/graphic.cpp:618]: I think you need to delete fakepoints here too as MCU_round_rect I believe allocates a new array if the input array is nil.

**Comment by runrevmark** [comment]
Hi Mark - this is almost there I think... There's a potential memory leak I've marked in the pull request, but apart from that it looks fine.

In terms of the oval / arc etc. then this is a little tricky, yes - I'll have a think about the best way to handle that (for completeness it would be useful to have).

Separating out the point computation methods has made the code a bit neater (as you mentioned above MCGraphic is not the prettiest of coding constructs!).

Thanks,

Mark.

**Comment by mwieder** [comment]
Ah... good point. I guess it makes sense that the point array has to be created _somewhere_. I'll get that patched up tonight. Also - do you have a preference for coding style between CamelCase and snake_case? I notice that most of the functions are just stringwordstogetherstyle which I find hard to read.

**Comment by runrevmark** [comment]
Hi Mark - thanks for getting that sorted quickly. I've merged this into develop via 'feature-effective_graphic_points' in the livecode repo. This should appear in the next dp of 6.7.

Thanks again,

Mark.

**Pull request closed without merging**